_source/_posts/2019-10-28-simple-secure-authentication-with-codeigniter.md

```diff
@@ -0,0 +1,843 @@
+                        <p>
+                            Logged in as <?php echo $username; ?>
+                        </p>
+                        <p>
+                            <a href="<?php echo site_url('news/create'); ?>">New article</a>
+                            | <a href="<?php echo site_url('logout'); ?>">Log Out</a>
+                        </p>
+                <?php
+                    } else {
+                ?>
+                        <p>Not logged in</p>
+                        <p><a href="<?php echo site_url('login'); ?>">Log In</a></p>
+                <?php
+                    }
+                ?>
+```
+
+Modify the News controller constructor to pass the logged-in user data to the views:
+
+`application/controllers/News.php`
+
+```php
+...
+    public function __construct()
+    {
+        parent::__construct();
+        $this->load->helper('url_helper');
+
+        $this->username = $this->session->username ?? null;
+        $data['username'] = $this->username;
+        $this->load->view('templates/header', $data);
+    }
+...
+```
+
+Reload the site and you'll see the navigation row. Test the login/logout links and confirm they are working correctly. You should be able to log in now and see your email address in the navigation bar when logged in.
+
+## Add a Form to Create News Items
+
+If you look carefully, you'll notice that when the user is logged in, the menu shows an additional item - **New article**. However, it's not implemented yet and will result in an error if you click it. You'll implement the new article form in this section of the tutorial.
+
+Modify the News controller and add the method to handle this action:
+
+`application/controllers/News.php`
+
+```php
+...
+    public function __construct()
+    {
+        parent::__construct();
+        $this->load->model('news_model');
+        $this->load->helper('url_helper');
+
+        $this->username = $this->session->username ?? null;
+        $data['username'] = $this->username;
+        $this->load->view('templates/header', $data);
+    }
+
+...
+
+    public function create()
+    {
+        if (! $this->username) {
+            redirect('login');
+        }
+
+        $this->load->helper('form');
+        $this->load->library('form_validation');
+
+        $data['title'] = 'Create a news item';
+
+        $this->form_validation->set_rules('title', 'Title', 'required');
+        $this->form_validation->set_rules('text', 'Text', 'required');
+
+        if ($this->form_validation->run() === FALSE)
+        {
+            $this->load->view('news/create', $data);
+        } else {
+            $this->news_model->set_news();
+            $this->load->view('news/success');
+        }
+
+        $this->load->view('templates/footer');
+    }
+```
+
+Note how the route is guarded by checking if there is a logged-in user before allowing the code to execute:
+
+```php
+        if (! $this->username) {
+            redirect('login');
+        }
+```
```

Comment:
I would left-align this code block.

Reply:
this indentation matches the code in the file so I like to leave it as is